server/src-lib/Hasura/RQL/IR/OrderBy.hs

```diff
@@ -111,52 +62,21 @@ orderByColFromTxt =
 
 -- order by item
 
-data OrderByItemG a
+data OrderByItemG (b :: Backend) a
   = OrderByItemG
-  { obiType   :: !(Maybe OrderType)
+  { obiType   :: !(Maybe (BasicOrderType b))
   , obiColumn :: !a
-  , obiNulls  :: !(Maybe NullsOrder)
-  } deriving (Show, Eq, Lift, Functor, Foldable, Traversable, Generic)
-instance (Hashable a) => Hashable (OrderByItemG a)
-
-type OrderByItem = OrderByItemG OrderByCol
-
-$(deriveToJSON (aesonDrop 3 snakeCase){omitNothingFields=True} ''OrderByItemG)
-
--- Can either be string / object
-instance FromJSON OrderByItem where
-  parseJSON (String t) =
-    case Atto.parseOnly orderByParser t of
-    Right r -> return r
-    Left _  ->
-      fail "string format for 'order_by' entry : {+/-}column Eg : +posted"
-
-  parseJSON (Object o) =
-    OrderByItemG
-    <$> o .:? "type"
-    <*> o .:  "column"
-    <*> o .:? "nulls"
-  parseJSON _ = fail "expecting an object or string for order by"
-
-newtype OrderByExp
-  = OrderByExp { getOrderByItems :: [OrderByItem] }
-  deriving (Show, Eq, ToJSON, Lift)
-
-instance FromJSON OrderByExp where
-  parseJSON v@(String _) =
-    OrderByExp . (:[]) <$> parseJSON v
-  parseJSON v@(Array _) =
-    OrderByExp <$> parseJSON v
-  parseJSON v@(Object _) =
-    OrderByExp . (:[]) <$> parseJSON v
-  parseJSON _ =
-    fail "Expecting : array/string/object"
-
-orderByParser :: AttoT.Parser Text OrderByItem
-orderByParser =
-  OrderByItemG <$> otP <*> colP <*> return Nothing
-  where
-    otP  = ("+" *> return (Just $ OrderType S.OTAsc))
-           <|> ("-" *> return (Just $ OrderType S.OTDesc))
-           <|> return Nothing
-    colP = Atto.takeText >>= orderByColFromTxt
+  , obiNulls  :: !(Maybe (NullsOrderType b))
+  } deriving (Functor, Foldable, Traversable, Generic)
+deriving instance (Representation b, Show a) => Show (OrderByItemG b a)
+deriving instance (Representation b, Eq a)   => Eq   (OrderByItemG b a)
+deriving instance (Representation b, Lift a) => Lift (OrderByItemG b a)
+instance (Representation b, Hashable a) => Hashable (OrderByItemG b a)
+
+type OrderByItem b = OrderByItemG b OrderByCol
+
+instance (Representation b, FromJSON a) => FromJSON (OrderByItemG b a) where
+  parseJSON = genericParseJSON (aesonDrop 3 snakeCase){omitNothingFields=True}
+
+instance ToJSON a => ToJSON (OrderByItemG 'Postgres a) where
+  toJSON = genericToJSON (aesonDrop 3 snakeCase){omitNothingFields=True}
```

Comment:
I like the pattern that the files in `Hasura.RQL` are assuming with these changes.